src/lsm/cache_map.zig

```diff
@@ -0,0 +1,358 @@
+            try map_2.ensureTotalCapacity(allocator, options.map_value_count_max);
+            errdefer map_2.deinit(allocator);
+
+            var scope_map: Map = .{};
+            try scope_map.ensureTotalCapacity(allocator, options.scope_value_count_max);
+            errdefer scope_map.deinit(allocator);
+
+            return Self{
+                .cache = cache,
+                .map_1 = map_1,
+                .map_2 = map_2,
+                .scope_map = scope_map,
+                .options = options,
+            };
+        }
+
+        pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
+            self.cache.deinit(allocator);
+            self.map_1.deinit(allocator);
+            self.map_2.deinit(allocator);
+            self.scope_map.deinit(allocator);
+        }
+
+        pub inline fn has(self: *Self, key: Key) bool {
+            return self.cache.get_index(key) != null or
+                self.map_1.getKeyPtr(tombstone_from_key(key)) != null or
+                self.map_2.getKeyPtr(tombstone_from_key(key)) != null;
+        }
+
+        pub inline fn get(self: *Self, key: Key) ?*Value {
+            return self.cache.get(key) orelse
+                self.map_1.getKeyPtr(tombstone_from_key(key)) orelse
+                self.map_2.getKeyPtr(tombstone_from_key(key));
+        }
+
+        pub fn upsert(self: *Self, value: *const Value) void {
+            self.last_upsert_was_update_with_eviction = false;
+            _ = self.cache.upsert_index(value, on_eviction);
+
+            if (self.scope_is_active and !self.last_upsert_was_update_with_eviction) {
+                if (self.map_1.getKey(value.*)) |stash_value| {
+                    // Scope Map: Case 2a.
+                    _ = self.scope_map.putAssumeCapacity(stash_value, {});
+                } else {
+                    // Scope Map: Case 2b.
+                    _ = self.scope_map.putAssumeCapacity(
+                        tombstone_from_key(key_from_value(value)),
+                        {},
+                    );
+                }
+            }
+        }
+
+        fn on_eviction(cache: *Cache, value: *const Value, updated: bool) void {
+            var self = @fieldParentPtr(Self, "cache", cache);
+            if (updated) {
+                // Scope Map: Case 1.
+                self.last_upsert_was_update_with_eviction = true;
+                if (self.scope_is_active) {
+                    _ = self.scope_map.putAssumeCapacity(value.*, {});
+                }
+            } else {
+                if (self.scope_is_active) {
+                    // Scope Map: Case 3.
+                    _ = self.scope_map.putAssumeCapacity(value.*, {});
+
+                    const gop = self.map_1.getOrPutAssumeCapacity(value.*);
+                    gop.key_ptr.* = value.*;
+                } else {
+                    const gop = self.map_1.getOrPutAssumeCapacity(value.*);
+                    gop.key_ptr.* = value.*;
```

Comment:
Nope - the `scope_map` is explicitly a `putAssumeCapacity` so it'll only `put` the first time (given the confusing behavior of `put` when updating keys - since the first put is what we want to rollback to) but the `map_1` needs to always be the latest value.

Reply:
Ah, it took me a bit to understand; `put` calls `getOrPut` but it only overwrites `gop.value_ptr` while we update `gop.key_ptr` so subsequent `put` don't actually overwrite the value in our case.

Reply:
I should definitely explain this better with a comment; in fact, originally there was one somewhere else saying as much but I didn't understand it as written and took a bit of head banging to realize :smile: